File: parley/src/main/java/nu/parley/android/data/model/MimeType.java
package nu.parley.android.data.model;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public enum MimeType {
    ImageJpeg("image/jpeg"),
    ImagePng("image/png"),
    ImageGif("image/gif"),
    ApplicationPdf("application/pdf"),
    Unknown("unknown");

    static MimeType from(String key) {
        for (MimeType mimeType : values()) {
            if (mimeType.key.equals(key)) {
                return mimeType;
            }
        }
        return Unknown;
    }

    public static List<MimeType> getImages() {
        return new ArrayList<>(Arrays.asList(
                MimeType.ImageJpeg,
                MimeType.ImagePng,
                MimeType.ImageGif
        ));
    }

    public static List<MimeType> getDocuments() {
        return new ArrayList<>(Collections.singletonList(
                MimeType.ApplicationPdf
        ));
    }

    public final String key;

    MimeType(String key) {
        this.key = key;
    }
}

File: parley/src/main/java/nu/parley/android/data/model/MimeType.kt
package nu.parley.android.data.model

import java.io.File

enum class MimeType(
    val value: String,
    val extensions: List<String>,
) {
    ImageJpeg("image/jpeg", listOf("jpg", "jpeg")),
    ImagePng("image/png", listOf("png")),
    ImageGif("image/gif", listOf("gif")),
    ApplicationPdf("application/pdf", listOf("pdf")),
    Other("*/*", listOf());

    fun isImage() = when (this) {
        ImageJpeg,
        ImagePng,
        ImageGif -> true

        ApplicationPdf,
        Other -> false
    }

    fun isDocument() = when (this) {
        ImageJpeg,
        ImagePng,
        ImageGif -> false

        ApplicationPdf -> true
        Other -> false
    }

    fun isFile() = when (this) {
        ImageJpeg,
        ImagePng,
        ImageGif -> false

        ApplicationPdf,
        Other -> true
    }

    fun getExtension() = extensions.first()

    companion object {
        fun fromUrl(url: String) = fromExtension(File(url).extension)
        fun fromValue(value: String) = entries.firstOrNull { it.value == value } ?: Other
        fun fromExtension(extension: String) =
            entries.firstOrNull { it.extensions.contains(extension) } ?: Other

        val images = entries.filter { it.isImage() }
        val documents = entries.filter { it.isDocument() }
    }
}
